### ISSUE TITLE
Assertion failure in SelectionDAG

### ISSUE THREAD

**Issue by ilovepi** [llvm/llvm-project#58675]
We're seeing an assertion failure in SelectionDAG when building Fuchsia

Failing Bot:
https://ci.chromium.org/ui/p/fuchsia/builders/ci/clang_toolchain.ci.core.arm64-release/b8799099192459783297/overview

Diagnostic output:
```console
[89930/276809](64) CXX obj/src/media/audio/lib/timeline/timeline.timeline_rate.cc.o
FAILED: obj/src/media/audio/lib/timeline/timeline.timeline_rate.cc.o
../../../recipe_cleanup/clangxe2ptd3r/bin/clang++ -MD -MF obj/src/media/audio/lib/timeline/timeline.timeline_rate.cc.o.d -D_LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS -D_LIBCPP_REMOVE_TRANSITIVE_INCLUDE...
clang++: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:10274: void llvm::SelectionDAG::ReplaceAllUsesWith(SDNode *, SDNode *): Assertion `(!From->hasAnyUseOfValue(i) || From->getValueType(i) == To...
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: ../../../recipe_cleanup/clangxe2ptd3r/bin/clang++ -MD -MF obj/src/media/audio/lib/timeline/timeline.timeline_rate.cc.o.d -D_LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS -D_LIBCPP_REM...
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module '../../src/media/audio/lib/timeline/timeline_rate.cc'.
4.	Running pass 'AArch64 Instruction Selection' on function '@_ZN5media12TimelineRate7ProductES0_S0_b'
#0 0x0000564835ef2144 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (../../../recipe_cleanup/clangxe2ptd3r/bin/clang+++0x3fbb144)
clang++: error: clang frontend command failed with exit code 134 (use -v to see invocation)
Fuchsia clang version 16.0.0 (https://llvm.googlesource.com/llvm-project 7aa0968842c5c36595042257009d1dc1a244ff6a)
Target: aarch64-unknown-fuchsia
Thread model: posix
InstalledDir: ../../../recipe_cleanup/clangxe2ptd3r/bin
clang++: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang++: note: diagnostic msg: clang-crashreports/timeline_rate-37a32c.cpp
clang++: note: diagnostic msg: clang-crashreports/timeline_rate-37a32c.sh
clang++: note: diagnostic msg:

********************
```
Reproducers:
[clang-crashreports.zip](https://github.com/llvm/llvm-project/files/9890819/clang-crashreports.zip)

I've bisected this down to https://reviews.llvm.org/D127531 in https://reviews.llvm.org/rG2ec51f1c753a63c786805708eb7875086cc5da9d

**Comment by ilovepi** [comment]
I forget that this is the better way to link a revision in github: 2ec51f1c753a63c786805708eb7875086cc5da9d

I've also notified the authors on D127531 directly.

**Comment by ilovepi** [comment]
I think I grabbed the wrong commit from my bisect earlier. I believe this is related to 01ff511593d1. Reverting locally stops the reproducer from triggering the assert.

**Comment by vfdff** [comment]
Thanks for your report, I'll take a look

**Comment by vfdff** [comment]
reduced case: llc -mtriple=aarch64-linux-gnu -O2 reduced.ll
```
define [2 x i64] @foo(i128 %a.addr, i128 %b.addr) {
entry:
  br label %do.body

do.body:                                      ; preds = %do.body, %entry
  %a.addr.i1 = phi i128 [ 1, %do.body ], [ 0, %entry ]
  %b.addr.i2 = phi i128 [ %sub, %do.body ], [ 0, %entry ]
  %0 = tail call i128 @llvm.umin.i128(i128 %a.addr, i128 %b.addr.i2)
  %1 = tail call i128 @llvm.umax.i128(i128 0, i128 %a.addr)
  %sub = sub i128 %b.addr, %0
  %cmp18.not = icmp eq i128 %b.addr, %0
  br i1 %cmp18.not, label %do.end, label %do.body

do.end:                                       ; preds = %do.body
  ret [2 x i64] zeroinitializer
}

declare i128 @llvm.umin.i128(i128, i128)
declare i128 @llvm.umax.i128(i128, i128)
```

**Comment by vfdff** [comment]
failed because the type mismatch of **From->getValueType(1) == To->getValueType(1)** in SelectionDAG::ReplaceAllUsesWith
```
(gdb) p From->dump()
t58: i64,glue = <<Unknown Node #482>> t15, t77
$16 = void
(gdb) p To->dump()
t72: i64,i32 = <<Unknown Machine Node #4294966369>> t15, t77
```

**Comment by ilovepi** [comment]
This seems to be fixed in 63a46385f2c6dd39cf68d9811548c53e8d460cd9 from what I can see.
